### ISSUE TITLE
Permit: order hooks mismatch

### ISSUE THREAD

**Issue by alfetopito** [cowprotocol/cowswap#3400]
**Description**

Researching reported issue in [this thread](https://cowservices.slack.com/archives/C0361CDG8GP/p1699889463701179)
Which has been reported o Discord and appzi

Checking for the reported error (Insufficient fee) in the backend logs in the past 24h

64 occurrences:

https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-24h%2Fh,to:now))&_a=(columns:!(kubernetes.pod_name,log),filters:!(),index:c0e240e0-d9b3-11ed-b0e6-e361adffce0b,interval:auto,query:(language:kuery,query:'InsufficientFee%20and%20calldata'),sort:!(!('@timestamp',desc)))

Checking the latest for quote `294643412`

https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-24h%2Fh,to:now))&_a=(columns:!(kubernetes.pod_name,log),filters:!(),index:c0e240e0-d9b3-11ed-b0e6-e361adffce0b,interval:auto,query:(language:kuery,query:'294643412'),sort:!(!('@timestamp',desc)))

Quote did not contain permit hook, thus order placement - which did contain it - rightfully failed

Other users/quotes though, did include the hook in the quote, but also failed due to low gas estimation
This was faced during development and not 100% mitigated

Exacmple https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-24h%2Fh,to:now))&_a=(columns:!(kubernetes.pod_name,log),filters:!(),index:c0e240e0-d9b3-11ed-b0e6-e361adffce0b,interval:auto,query:(language:kuery,query:'294622345'),sort:!(!('@timestamp',desc)))

Quote

```
2023-11-15T15:43:59.487Z DEBUG request{id="ce8b06c3fd4d076770580a9f1afc7632"}: shared::order_quoting: finished computing quote response=OrderQuoteResponse { quote: OrderQuote { sell_token: 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48, buy_token: 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee, receiver: Some(0xed42e2cb468df0106874b6629de66cbcd7dd13fe), sell_amount: 949326684, buy_amount: 471824082958811630, valid_to: 1700064836, app_data: Both { full: "{\"appCode\":\"CoW Swap\",\"environment\":\"production\",\"metadata\":{\"hooks\":{\"pre\":[{\"callData\":\"0xd505accf000000000000000000000000ff65d1dfcf256cf4a8d5f2fb8e70f936606b7474000000000000000000000000c92e8bdf79f0507f65a392b0ab4667716bfe0110ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000000000000000000000000000000000006ebc8b24000000000000000000000000000000000000000000000000000000000000001b4a571a71da8235440f49a0a794cc5cb163bb2c73eaa727439a5a13987295f36c723bc4e45ee55b7e4757a11b14a228ad974260462732d09bc750b9f23156046c\",\"gasLimit\":\"80000\",\"target\":\"0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48\"}]},\"orderClass\":{\"orderClass\":\"market\"},\"quote\":{\"slippageBips\":\"50\"},\"utm\":{\"utmContent\":\"header-cta-button\",\"utmMedium\":\"web\",\"utmSource\":\"cow.fi\"}},\"version\":\"0.11.0\"}", expected: 0x4a453fb63b41cea1ef6bd2e7e2de01dd1dd1a0e52f498fb599d78c7b32b40df9 }, fee_amount: 50673316, kind: Sell, partially_fillable: false, sell_token_balance: Erc20, buy_token_balance: Erc20, signing_scheme: Eip712 }, from: 0xed42e2cb468df0106874b6629de66cbcd7dd13fe, expiration: 2023-11-15T15:45:57.756715517Z, id: Some(294622345) }
```

Order
```
2023-11-15T15:45:12.227Z DEBUG request{id="d2c8cedf37b57d895ef76f1b689ecc08"}: orderbook::api::post_order: error creating order order=OrderCreation { sell_token: 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48, buy_token: 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee, receiver: Some(0xed42e2cb468df0106874b6629de66cbcd7dd13fe), sell_amount: 949326684, buy_amount: 469464962544017571, valid_to: 1700064840, fee_amount: 50673316, kind: Sell, partially_fillable: false, sell_token_balance: Erc20, buy_token_balance: Erc20, from: Some(0xed42e2cb468df0106874b6629de66cbcd7dd13fe), signature: Eip712("0x0ab7d8899f19bc998857a2613cd84ea0b1e46ce932a90a322f72a2fdf60a908f0c85de8a6feec850d31457fae30f9bc19d961dc7fce6c5f8a45b2a278640ba9a1c"), quote_id: Some(294622345), app_data: Both { full: "{\"appCode\":\"CoW Swap\",\"environment\":\"production\",\"metadata\":{\"hooks\":{\"pre\":[{\"callData\":\"0xd505accf000000000000000000000000ed42e2cb468df0106874b6629de66cbcd7dd13fe000000000000000000000000c92e8bdf79f0507f65a392b0ab4667716bfe0110ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000000000000000000000000000000000006ebc8ea5000000000000000000000000000000000000000000000000000000000000001cf60304d62bcaa36de4c61b0fc5ae92da60b4fb0293016cfb4222050ada17534e49a8807c04e720de9d34743c9c5e071836e47ef192cac6a2ba3621ff9435fde4\",\"gasLimit\":\"91497\",\"target\":\"0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48\"}]},\"orderClass\":{\"orderClass\":\"market\"},\"quote\":{\"slippageBips\":\"50\"},\"utm\":{\"utmContent\":\"header-cta-button\",\"utmMedium\":\"web\",\"utmSource\":\"cow.fi\"}},\"version\":\"0.11.0\"}", expected: 0x1de3fe360ad66cef1166b3860d1308c5451ba6696bae6cebd62c8a0a01315876 } } err=OrderValidation(InsufficientFee)
```

I suspect the issue for the second case is that the account used for signing the quote requests had approvals executed on its behalf https://etherscan.io/tokenapprovalchecker?search=0xff65d1dfcf256cf4a8d5f2fb8e70f936606b7474

That is bad because once approved, subsequent approvals become cheaper gas wise. See this simulation https://www.tdly.co/shared/simulation/e6f0776a-b660-4183-9842-a48778d69402
Then, we end up with values too low, which can't be matched for first time approvers

How is this account which has no trades getting approvals executed, then?
In cow swap execution, of course https://etherscan.io/tx/0x142b7bc65193606befe25a5b374448e132e5daed36f8aab707e302a2590bb5c0#eventlog

The permit was part of this order https://explorer.cow.fi/orders/0x5063d6ee80939076c0749399284ea109a5e340ce324bfdd68bee8a0777fde370aa58bf8343e0e84f0cb9b6ff211d60f49c16aad36554b649?tab=overview
It simulates successfully, with the caveat that it's `from` the internal address (`0xff65d1dfcf256cf4a8d5f2fb8e70f936606b7474`), not the order owner's (`0xaa58bf8343e0e84f0cb9b6ff211d60f49c16aad3`) https://www.tdly.co/shared/simulation/dba8eb77-10a0-43ac-95fa-92df97be21a1

So the user paid gas for executing the approval of the internal static account the shouldn't have any approvals.

Which might be due to the same reason as the first problem.
Hooks are being included in orders when they shouldn't be

**How to Reproduce**
Unclear, more experimentation needed

**Expected behavior**
Hooks should only be included when needed, and matching the order being placed

**Comment by alfetopito** [comment]
Continuing the analysis, I can see that it the errors became more frequent from Nov-10th
![image](https://github.com/cowprotocol/cowswap/assets/43217/a3a23c56-4d26-4a44-96de-1198af9232b9)

That coincides somewhat with [1.49.0](https://github.com/cowprotocol/cowswap/releases/tag/v1.49.0) release.

Not yet sure whether there was a bug introduced causing the issue to be more frequent (keep in mind it was happening before but much less frequent) or whether permits became more common now that the permittable tokens are flagged in the token list dropdown.

Looking for analytics data to verify the usage of permit and confirm the suspicion.

**Comment by alfetopito** [comment]
[Dune query](https://dune.com/queries/3136513/5230767?blockchains=ethereum&namespace=cow_protocol&category=abstraction&table=trades)

The number of gasless orders increased, but not as much to indicate the same levels of errors as before.
![image](https://github.com/cowprotocol/cowswap/assets/43217/1343cfd1-c197-40dc-9a9a-eb1a71ae7fef)

It looks like some changes introduced caused the failures to be more frequent.

**Comment by alfetopito** [comment]
One more important information I just noticed.
All the quote account approvals were done from orders originated in a SafeApp.
Which shouldn't have hook data at all!

- [OX](https://explorer.cow.fi/orders/0x5063d6ee80939076c0749399284ea109a5e340ce324bfdd68bee8a0777fde370aa58bf8343e0e84f0cb9b6ff211d60f49c16aad36554b649?tab=overview)
- [USDC](https://explorer.cow.fi/orders/0x064198c7c3472ebce172c74e7888363734b39bd7ffcd8837bb176443962b7e919f2f472b1276677c3c7e264c2ec125430e2f27af6541020a?tab=overview)
- [COW](https://explorer.cow.fi/orders/0x18e618978ba8768639408726ef2f43b8f3d4eac87fa127f3b3755686c02a7d8e646dd8d94e1e46d24f649d343655f34a4d63d88665559ab8?tab=overview)
- [DAI](https://explorer.cow.fi/orders/0x2f79c05038d75e87b8cd632e7b8f9b34689743c0d181c12317e58bdd21f6dd31aa58bf8343e0e84f0cb9b6ff211d60f49c16aad36554b796?tab=overview)
- [wstETH](https://explorer.cow.fi/orders/0xb4e50bf5772bfb8caad7ed4a5337f645c1c2c1def42a3068cc4bd0347e9e524390dd9d99770fbdf2dea5cf367861e5eb41cfd22265517a1e?tab=overview)
- [ENS](https://explorer.cow.fi/orders/0x5ca2ee1ddd1b38e040ce86dfc3891dc775c8900ee3d9fe28d246d64ed29613dfd90ff04a87e7b0cf01c83618752933ca15aa49ae654e4828?tab=overview)

**Comment by alfetopito** [comment]
Another problem: TWAP orders are having calldata included in them. They SHOULD NOT!!

[logs query](
https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-30d%2Fd,to:now))&_a=(columns:!(kubernetes.pod_name,log),filters:!(),index:c0e240e0-d9b3-11ed-b0e6-e361adffce0b,interval:auto,query:(language:kuery,query:'calldata%20and%20twap'),sort:!(!('@timestamp',desc))))

![image](https://github.com/cowprotocol/cowswap/assets/43217/1cd3e85c-574f-4284-9dde-7f4ffebd98b5)

**Comment by alfetopito** [comment]
Although the previous query looks bad and shouldn't happen, it was mostly in the quotes.

[Here](https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-30d%2Fd,to:now))&_a=(columns:!(kubernetes.pod_name,log),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:c0e240e0-d9b3-11ed-b0e6-e361adffce0b,key:kubernetes.pod_name,negate:!t,params:(query:mainnet-solver-prod),type:phrase),query:(match_phrase:(kubernetes.pod_name:mainnet-solver-prod))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:c0e240e0-d9b3-11ed-b0e6-e361adffce0b,key:kubernetes.pod_name,negate:!t,params:(query:autopilot-prod),type:phrase),query:(match_phrase:(kubernetes.pod_name:autopilot-prod)))),index:c0e240e0-d9b3-11ed-b0e6-e361adffce0b,interval:auto,query:(language:kuery,query:'calldata%20and%20twap%20and%20not%20order_quoting%20and%20not%20log:%20post_quote'),sort:!(!('@timestamp',desc)))) are the actual TWAP orders having hooks included:

![image](https://github.com/cowprotocol/cowswap/assets/43217/07c64325-d7a4-4e68-8789-b56697da329d)

**Comment by alfetopito** [comment]
Fixed on #3404
